net/tstun/linkattrs_linux.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tstun
+
+import (
+	"fmt"
+	"unsafe"
+
+	"golang.zx2c4.com/wireguard/tun"
+
+	"golang.org/x/sys/unix"
+)
+
+// ethtoolCmd is struct ethtool_cmd from linux/ethtool.h
```

Comment:
```
 * struct ethtool_cmd - DEPRECATED, link control and status                                                                                                                                                               
 * This structure is DEPRECATED, please use struct ethtool_link_settings.                                                                                                                                                 
```

Did you try using `ethtool_link_settings`?

Reply:
Yes to this. That didn't work at least on my machine, in the sense that the GLINKSETTINGS cmd returns 0 but fills the struct with speed = 0, and submitting a SLINKSETTINGS fails with EINVAL. The recommended way is to try GLINKSETTINGS first and then fall back, which is probably the right thing to do after all, but the silent failure (?) is unfortunate. I'll investigate more while the x/sys/unix change is pending.

Reply:
Ah, I remembered a suspected cause and it checks out: per docs, speed = 0 can be observed when the link is down, and presumably one cannot modify it in that state either. This doesn't go away when the TUN device observes EventUp; it probably has to be done via the link monitor or by implementing `waitForInterfaceUp` on linux. No such caveat is given for ethtool_cmd, and, in fact, it appears to work just fine where it is.